STYLEGUIDE.adoc

```diff
@@ -0,0 +1,426 @@
+
+The contents of this document are “recommendations”, that is:
+They are more concrete than “principles” (like “write simple code”) because we want them to be clearly action-guiding.
+They are more vague than “rules” (like “never use function X”) because it’s usually impossible to be that precise, and so judgement is still required in applying them.
+
+== Haskell
+
+=== Basic formatting
+
+Use 4 spaces for indentation. Avoid lines over 120 characters. (This is codified in `.editorconfig`.)
+
+Format Haskell source with `stylish-haskell` (config in `.stylish-haskell.yaml`).
+
+==== Reasoning
+
+The indentation is arbitrary. The line length is longer than the “traditional” 80 characters, but anecdotally Haskell code tends toward longer lines, and breaking them up doesn’t help readability much.
+
+`stylish-haskell` is a fairly modest formatter that doesn’t touch too much. At the moment we mostly use it for formatting imports!
+
+In the past we tried `ormolu`, but its https://github.com/tweag/ormolu/issues/641[lack of support for multi-line comments] was a deal-breaker.
+
+=== Write type signatures
+
+Every top-level binding should have a type signature. _Most_ other named bindings (e.g. `let` and `where` bindings) should also have type signatures.
+
+==== Reasoning
+
+Haskell is traditionally lauded for its excellent type inference. People used to make the argument that this brought Haskell closer to dynamically-typed languages in ease of use: you don't have to write type signatures, the compiler will infer it for you.
+
+However, not writing type signatures has a heavy _maintainability_ cost.
+
+.1. Non-local errors
+
+If types are inferred, then the way a binding is _used_ can affect the type which is inferred for it. That means that a mistake in using a binding can result in an error _inside_ the binding (or elsewhere) due to the inferred type not matching what goes on in the RHS of the binding. More generally, errors can end up appearing in unexpected and counter-intuitive places.
+
+Pinning down the type of a binding means that any errors relating to using that binding will occur at the use site, where they belong!
+
+.2. Documentation
+
+It's a Haskell truism that the type forms part of the documentation of a binding. But that requires you to be able to _see_ the type. Of course, in this day and age, we should all have an IDE that shows us the type on hover. But sometimes you're stuck using `vim`. Or you're reviewing the code on Github. Or the IDE is broken.
+
+So do your colleagues a favor and just write it down.
+
+.3. Silent changes
+
+Changing the type of a binding often means that something relatively significant has changed. But if the type is inferred, this can happen without you noticing it. This is almost always bad!
+
+=== Write kind signatures
+
+Every type definition that has parameters which are not all of kind `Type` should have a kind signature using `StandaloneKindSignatures`.
+
+==== Examples
+Bad:
+```
```

Comment:
Asciidoc has source blocks with specific languages, which in principle should get us syntax higlighting: https://docs.asciidoctor.org/asciidoc/latest/verbatim/source-blocks/